### PR TITLE
Allow users to request referenced resources along with the searched via inline params.

### LINKED ISSUE #1947: Implement _revinclude on Search API
**Is your feature request related to a problem? Please describe.**
Implement `_revinclude` on the client Search API. This should cause the search response to include resources (that have a relationship to the primary resources) in the returned result.

**Describe the solution you'd like**
Implement a new function (`Search#include`) like the `Search#has` function in the search DSL that should cause the search response to include resources that have a relationship with the primary resource in the returned search result.

**Describe alternatives you've considered**
None that is related to using existing functionality on the search DSL. I however contemplated running two queries i.e. run a query to fetch the primary resources, then run a custom SQL query (**not supported**) that uses SQL UNION  against the `ReferenceIndexEntity` table to fetch the accompanying resources for each of the primary resource. The example below fetches accompanying resources (Tasks, Conditions and CarePlan) for the Patient referenced with id `Patient/fee261ad-4e90-4054-9cd3-906a60362c59`.

> NOTE: The custom query in this proposal CANNOT work with the current Search API as the functionality for running custom queries does not exist 

```sql
SELECT re.serializedResource
FROM ReferenceIndexEntity ref
INNER JOIN ResourceEntity re on ref.resourceUuid = re.resourceUuid
WHERE ref.resourceType = 'Task'
  AND index_name = 'subject'
  AND index_value = 'Patient/fee261ad-4e90-4054-9cd3-906a60362c59'
UNION  ALL
SELECT re.serializedResource
FROM ReferenceIndexEntity ref
INNER JOIN ResourceEntity re on ref.resourceUuid = re.resourceUuid
WHERE ref.resourceType = 'Condition'
  AND index_name = 'subject'
  AND index_value = 'Patient/fee261ad-4e90-4054-9cd3-906a60362c59'
UNION  ALL
SELECT re.serializedResource
FROM ReferenceIndexEntity ref
INNER JOIN ResourceEntity re on ref.resourceUuid = re.resourceUuid
WHERE ref.resourceType = 'CarePlan'
  AND index_name = 'subject'
  AND index_value = 'Patient/fee261ad-4e90-4054-9cd3-906a60362c59';
```

**Additional context**
The implementation for this function should be close to `Search#has` for instance in this example:
_Fetch all Patients with Condition Diabetes_, `Search#has` would return just the Patient resources with the Condition, whereas the new function should also include the `Conditions` in the result.
See [Search parameter's _revinclude](https://www.hl7.org/fhir/search.html#_revinclude)

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1947

**Description**
1. Add revInclude function to search api's w.r.t https://build.fhir.org/search.html#_revinclude .

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one:  Feature 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt [hunk 1]
```diff
@@ -142,6 +143,27 @@ data class Search(val type: ResourceType, var count: Int? = null, var from: Int?
...
+   *
+   * ```
+   * fhirEngine.search<Resource>(Search(ResourceType.Patient).apply {
+   *  filter(Patient.GIVEN, { value = "James" })
+   *  revInclude(ResourceType.Encounter, Encounter.PATIENT)
+   *  revInclude(ResourceType.Condition, Condition.PATIENT)
+   * })
+   * ```
+   */
+  fun revInclude(resourceType: ResourceType, vararg clientParam: ReferenceClientParam) {
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt:164]: why is the client param `vararg`?

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt [hunk 2]
```diff
@@ -142,6 +143,27 @@ data class Search(val type: ResourceType, var count: Int? = null, var from: Int?
     sort = parameter
     this.order = order
   }
+
+  /**
+   * Allows user to include additional resources to be included in the search results that reference
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt:148]:
```suggestion
   * Includes additional resources in the search results that reference ...
```

File: engine/src/main/java/com/google/android/fhir/search/Search.kt [hunk 3]
```diff
@@ -35,3 +36,9 @@ suspend inline fun <reified R : Resource> FhirEngine.count(init: Search.() -> Un
 suspend fun FhirEngine.search(xFhirQuery: String): List<Resource> {
   return this.search(translate(xFhirQuery))
 }
+
+suspend fun FhirEngine.search(resourceType: ResourceType, init: Search.() -> Unit): List<Resource> {
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/search/Search.kt:40]: This weakens type safety by quite a bit.

can we do something like this: so you package the reverse included resources together with their parents?

```
fun <R, S> search(...) : List<Map<R, S>> { ... }
```

File: engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt [hunk 4]
```diff
@@ -166,6 +168,22 @@ internal fun Search.getQuery(
...
+        WHERE a.resourceType = ?
+        $filterStatement
+        $sortOrderStatement
+        $limitStatement
+        )
+        Select resourceUuid
+        FROM ResourceEntity
+        WHERE '${type.name}/' || resourceId in UUIDS ${revIncludeMap.toSQLQuery()}
+        )
+        """.trimIndent()
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt:185]: can you fix the indentation? all a bit hard to read for me.

File: engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt [hunk 5]
```diff
@@ -166,6 +168,22 @@ internal fun Search.getQuery(
         $limitStatement)
         """
         }
+        revIncludeMap.isNotEmpty() -> {
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt:171]: can you have nested and revinclude together? what happens in that case?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt (hunk 4 above)
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt:185]: can you try the annotation? @JSON... something like that..

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt [hunk 6]
```diff
@@ -53,6 +53,7 @@ data class Search(val type: ResourceType, var count: Int? = null, var from: Int?
   internal val uriFilterCriteria = mutableListOf<UriFilterCriteria>()
   internal var sort: IParam? = null
   internal var order: Order? = null
+  internal val revIncludeMap = mutableMapOf<ResourceType, MutableList<ReferenceClientParam>>()
```
[the remark this answers is not in the document]
Comment (ellykits) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt:56]: I think the iterate modifier will be a useful enhancement for the next iteration as well as an implementation of  **_include**  which is different from the **_revinclude** syntax.

[the remark this answers is not in the document]
Comment (ellykits) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt:56]: I think the iterate modifier will be a useful enhancement for the next iteration as well as an implementation of  **_include**  which is different from the **_revinclude** syntax.

[the remark this answers is not in the document]
Comment (ellykits) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt:56]: An example use case is : Retrieve a Group resource including all the referenced Resources of `Group#member` property.

File: engine/src/main/java/com/google/android/fhir/search/Search.kt (hunk 3 above)
Comment (ellykits) [reply engine/src/main/java/com/google/android/fhir/search/Search.kt:40]: I think `S` should instead be a `List<S>`

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt:56]: > For _revinlcude are we supporting just this syntax: `[_include|_revinclude]=[resource]:[parameter]` ? What about other syntaxes ?
> 
> ```
> [_include|_revinclude]=*
> [_include|_revinclude]=[resource]:*
> [_include|_revinclude]=[resource]:[parameter]:[targetType]
> ```

We don't necessarily want to implement the exact same spec as sever on the device. Instead we want to provide the subset of apis that would be useful and make most sense of the device. As the client app would only need a subset of resources,  I don't think we need to do the wildcard matching. 

Incase of [_include|_revinclude]=[resource]:[parameter]:[targetType], we can infer the [targetType] from the Resource on which the [_include|_revinclude] is being called upon.

File: engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt [hunk 7]
```diff
@@ -60,6 +62,29 @@ internal class FhirEngineImpl(private val database: Database, private val contex
...
+    val includedResources =
+      database.searchRev<R>(
+        search.getIncludeQuery(
+          includeIds = baseResources.map { "\'${it.resourceType}/${it.logicalId}\'" }
+        )
+      )
+    val resultMap = mutableMapOf<R, Map<ResourceType, List<Resource>>>()
+    baseResources.forEach { patient ->
+      resultMap[patient] =
+        includedResources
```
Comment (ellykits) [inline engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt:78]: We can convert this to a sequence using the `asSequence()` function.

**Review by ellykits — CHANGES_REQUESTED** [review]

**Comment by aditya-07** [comment]
I have added some experimental code in it for quick testing. Will convert it back once the changes are in better shape.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 8]
```diff
@@ -279,3 +285,14 @@ internal abstract class ResourceDao {
...
   }
 }
+
+internal data class IndexedIdAndSerializedResource(
+  @ColumnInfo(name = "index_value") val idOfBaseResourceOnWhichThisMatchedRev: String?,
+  @ColumnInfo(name = "resourceId") val idOfBaseResourceOnWhichThisMatchedInc: String?,
+  val serializedResource: String
+)
+
+internal data class IndexedIdAndResource(
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:304]: Done

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 9]
```diff
@@ -279,3 +285,14 @@ internal abstract class ResourceDao {
     }
   }
 }
+
+internal data class IndexedIdAndSerializedResource(
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:293]: Done

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt [hunk 10]
```diff
@@ -38,12 +38,10 @@ import com.google.android.fhir.search.filter.TokenParamFilterCriteria
 import com.google.android.fhir.search.filter.TokenParamFilterCriterion
 import com.google.android.fhir.search.filter.UriFilterCriteria
 import com.google.android.fhir.search.filter.UriParamFilterCriterion
-import org.hl7.fhir.r4.model.Patient
 import org.hl7.fhir.r4.model.ResourceType
 
 @SearchDslMarker
-data class Search(val type: ResourceType, var count: Int? = null, var from: Int? = null) {
-  internal val p = Patient()
+class Search(val type: ResourceType, var count: Int? = null, var from: Int? = null) : ISearch {
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt]: This change may pollute this PR with just the name change. So, planning to do it in a separate PR.

File: engine/src/main/java/com/google/android/fhir/search/ISearch.kt [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    operation: Operation = Operation.OR
+  )
+
+  fun sort(parameter: StringClientParam, order: Order)
+
+  fun sort(parameter: NumberClientParam, order: Order)
+
+  fun sort(parameter: DateClientParam, order: Order)
+
+  fun operations(operation: Operation)
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/ISearch.kt]: I have change it to properties instead and added docs to them.

File: engine/src/main/java/com/google/android/fhir/search/ISearch.kt [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import com.google.android.fhir.search.filter.QuantityParamFilterCriterion
+import com.google.android.fhir.search.filter.ReferenceParamFilterCriterion
+import com.google.android.fhir.search.filter.StringParamFilterCriterion
+import com.google.android.fhir.search.filter.TokenParamFilterCriterion
+import com.google.android.fhir.search.filter.UriParamFilterCriterion
+
+@DslMarker @Target(AnnotationTarget.CLASS, AnnotationTarget.TYPE) annotation class ISearchDsl
+
+@ISearchDsl
+interface ISearch {
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/ISearch.kt]: Done **BaseSearch**. Added documentation as well.

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 13]
```diff
@@ -127,3 +127,15 @@ suspend inline fun <reified R : Resource> FhirEngine.get(id: String): R {
 suspend inline fun <reified R : Resource> FhirEngine.delete(id: String) {
   delete(getResourceType(R::class.java), id)
 }
+
+typealias ReferencedResources = Map<ResourceType, List<Resource>>
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngine.kt]: Done. The referenced resources are of type `Map<ResourceType, Map<SearchParamName, List<Resource>>>`.

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt (hunk 1 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt]: Not applicable after refactor.

File: engine/src/main/java/com/google/android/fhir/search/Search.kt (hunk 3 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/Search.kt]: Done as part of refactor.

File: engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/SearchDsl.kt]: Done

File: engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt (hunk 5 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt]: Not applicable after refactor.

File: engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt (hunk 7 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt]: Not applicable after refactor.

File: engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt (hunk 4 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt]: Couldn't find a way to do it by annotations. Recomended spotless way is to use `spotless:off` and `spotless:on` [toggle](https://github.com/diffplug/spotless/tree/main/plugin-gradle#spotlessoff-and-spotlesson).

**Comment by aditya-07** [comment]
> I have added some experimental code in it for quick testing. Will convert it back once the changes are in better shape.

Experimental code has been removed or moved to a production ready state.

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 14]
```diff
@@ -3103,6 +3019,374 @@ class DatabaseImplTest {
       .inOrder()
   }
 
+  @Test
+  fun search_patient_with_reference_resources(): Unit = runBlocking {
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:3023]: can we also create smaller and more focused test cases.. for forward and backward includes.

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 15]
```diff
@@ -127,3 +127,17 @@ suspend inline fun <reified R : Resource> FhirEngine.get(id: String): R {
...
+typealias SearchParamName = String
+
+typealias ReferencedResources = Map<ResourceType, Map<SearchParamName, List<Resource>>>
+
+/** It contains the searched resource and referenced resources as per the search query. */
+data class SearchResult<R : Resource>(
+  /** Matching resource as per the query. */
+  val resource: R,
+  /** Matching referenced resources as per the [Search.include] criteria in the query. */
+  val included: ReferencedResources?,
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/FhirEngine.kt:140]: i think the forward include return type doesn't have to have resource type.

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt (hunk 14 above)
Comment (aditya-07) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:3217]: Done

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt (hunk 15 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngine.kt]: Done

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 16]
```diff
@@ -127,3 +127,50 @@ suspend inline fun <reified R : Resource> FhirEngine.get(id: String): R {
...
+      first.entries.zip(second.entries).all { (x, y) ->
+        x.key == y.key &&
+          x.value.size == y.value.size &&
+          x.value.zip(y.value).all { (x, y) -> equalsShallow(x, y) }
+      }
+    } else {
+      first?.size == second?.size
+    }
+
+  private fun equalsShallow2(
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/FhirEngine.kt:165]: avoid fuction names like this :)

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 17]
```diff
@@ -127,3 +127,50 @@ suspend inline fun <reified R : Resource> FhirEngine.get(id: String): R {
 suspend inline fun <reified R : Resource> FhirEngine.delete(id: String) {
   delete(getResourceType(R::class.java), id)
 }
+
+typealias SearchParamName = String
+
+/** It contains the searched resource and referenced resources as per the search query. */
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngine.kt]: Done

**Review by jingtang10 — APPROVED** [review]
great job!

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 18]
```diff
@@ -127,3 +127,54 @@ suspend inline fun <reified R : Resource> FhirEngine.get(id: String): R {
...
+ * Contains a FHIR resource that satisfies the search criteria in the query together with any
+ * referenced resources as specified in the query.
+ */
+data class SearchResult<R : Resource>(
+  /** Matching resource as per the query. */
+  val resource: R,
+  /** Matching referenced resources as per the [Search.include] criteria in the query. */
+  val included: Map<SearchParamName, List<Resource>>?,
+  /** Matching referenced resources as per the [Search.revInclude] criteria in the query. */
+  val revIncluded: Map<ResourceType, Map<SearchParamName, List<Resource>>>?
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngine.kt]: Done

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineTerminologyProvider.kt
Comment (jingtang10) [inline workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineTerminologyProvider.kt:1]: you could revert these changes i think

Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineTerminologyProvider.kt:1]: done

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt
[the remark this answers is not in the document]
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt:1]: done

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineDal.kt
[the remark this answers is not in the document]
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineDal.kt:1]: done

**Pull request merged**